Creds in header

ballad89 opened the pull request:
changes include
- support the recommended way of passing creds as specified in the [spec](https://tools.ietf.org/html/rfc6749#section-2.3.1). Newer authorizations servers expect creds in header not in form body
- ensure redirect_uri is consistent throughout the process, as stricter authorizations servers expect it to be exactly the same
- Added tests to make sure new parameter `creds-in-header?` is optional and does not break old tests

weavejester commented on README.md:
```diff
@@ -89,6 +89,11 @@ The landing URI is where the middleware redirects the user when the
 authentication process is complete. This could just be back to the
 index page, or it could be to the user's account page.
 
+* `:creds-in-header?`
```
`:basic-auth?` might be a better name for this option, since this option passes credentials through HTTP Basic Authentication.

weavejester commented on README.md:
```diff
@@ -89,6 +89,11 @@ The landing URI is where the middleware redirects the user when the
 
+* `:creds-in-header?`
+
+This is an optional parameter, which defaults to false. If set to true, it includes the client-id and secret as header
```
Lines should have same wrap length as rest of README.

weavejester commented on README.md:
```diff
@@ -89,6 +89,11 @@ The landing URI is where the middleware redirects the user when the
+* `:creds-in-header?`
+
+This is an optional parameter, which defaults to false. If set to true, it includes the client-id and secret as header
+`Authorization: Basic base64(id:secret)` as recommended by [spec](https://tools.ietf.org/html/rfc6749#section-2.3.1)
```
Instead try:

    as recommended by [the specification][].

    [the specification]: https://tools.ietf.org/html/rfc6749#section-2.3.1

weavejester commented on project.clj:
```diff
@@ -1,4 +1,4 @@
-(defproject ring-oauth2 "0.1.0"
+(defproject ring-oauth2 "0.2.0"
```
Please don't change the version number.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -41,16 +41,18 @@
+                       :form-params {:grant_type    "authorization_code"
+                                     :code          (get-in request [:query-params "code"])
+                                     :redirect_uri  (redirect-uri profile request)}}
+                 (true? creds-in-header?) (assoc :basic-auth [client-id client-secret])
```
Why do you use `true?` and `false?` here?

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -41,16 +41,18 @@
+                                     :redirect_uri  (redirect-uri profile request)}}
+                 (true? creds-in-header?) (assoc :basic-auth [client-id client-secret])
+                 (false? creds-in-header?) (assoc-in [:form-params :client_id] client-id)
+                 (false? creds-in-header?) (assoc-in [:form-params :client_secret] client-secret)))))
```
I think this would be better written:

```clojure
(http/post access-token-uri
           (-> {:accept :json, :as :json, :form-params (request-params request)}
               (cond-> header-credentials?
                 (add-header-credentials client-id client-secret)
                 (add-form-credentials   client-id client-secret))))
```

weavejester commented on test/ring/middleware/oauth2_test.clj:
```diff
@@ -90,3 +90,93 @@
+
+
+
+; Tests with new parameter
```
Remove whitespace and comment please.

weavejester commented on test/ring/middleware/oauth2_test.clj:
```diff
@@ -90,3 +90,93 @@
+
+; Tests with new parameter
+
+(def test-profile-2
```
As far as I can tell, you've just duplicated the functions and tests. Ideally we want one extra test, and we want to check to make sure the basic authentication header is transferred.

weavejester commented:
Thanks for the patch. I have a few review comments, and could you also fix the commit messages so that they're in line with the repository's standards?

ballad89 commented:
Cool... it's past midnight on my side, I'll have a look in the morning.

On Oct 15, 2017 23:52, "James Reeves" <notifications@github.com> wrote:

> Thanks for the patch. I have a few review comments, and could you also fix
> the commit messages so that they're in line with the repository's standards?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/weavejester/ring-oauth2/pull/5#issuecomment-336744305>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AELxNr-unGLYZJOCTPtk2oJfn7QVnlvtks5ssn6VgaJpZM4P55Zr>
> .
>

ballad89 commented:
I did to get rid of most of the intermediate commits, so I created a new branch and pull request off master and addressed the issues you raised
https://github.com/weavejester/ring-oauth2/pull/6
You can close this one

weavejester commented:
You don't need to create a new PR. It's preferable to fix the changes and push to the same branch.

ballad89 commented:
Yes, I wanted to rewrite the commit messages like you suggested, but a lot of the commits did not make sense in light of the changes .... but I understand your point
Would you want to continue with this branch or the new one ?

weavejester commented:
Whichever you prefer. If you've already opened a new PR, add a reference to it in this PR then close this one. In future though, I'd prefer that a proposed change is kept to a single PR, rather than the discussion split over two.

ballad89 commented:
I understand
Closing in preference to https://github.com/weavejester/ring-oauth2/pull/6
